src/Algebra/Construct/MonoidAction.agda

```diff
@@ -0,0 +1,227 @@
+open import Data.List.Relation.Binary.Pointwise as Pointwise
+  using (Pointwise; []; _∷_)
+import Data.List.Relation.Binary.Equality.Setoid as ≋
+open import Data.Product.Base using (_,_)
+
+open import Function.Base using (id; _∘_; flip)
+
+open import Level using (Level; suc; _⊔_)
+
+open import Relation.Binary.Bundles using (Setoid)
+open import Relation.Binary.Structures using (IsEquivalence)
+open import Relation.Binary.Core using (Rel)
+open import Relation.Binary.Definitions
+
+private
+  variable
+    a c r ℓ : Level
+    A : Set a
+    M : Set c
+
+
+------------------------------------------------------------------------
+-- Basic definitions: fix notation for underlying 'raw' actions
+
+module _ {c a ℓ r} {M : Set c} {A : Set a}
+         (_≈ᴹ_ : Rel M ℓ) (_≈_ : Rel A r)
+  where
+
+  private
+    variable
+      x y z : A
+      m n p q : M
+      ms ns ps qs : List M
+
+  record RawLeftAction : Set (a ⊔ r ⊔ c ⊔ ℓ)  where
+    infixr 5 _ᴹ∙ᴬ_ _ᴹ⋆ᴬ_
+
+    field
+      _ᴹ∙ᴬ_ : M → A → A
+      ∙-cong : m ≈ᴹ n → x ≈ y → (m ᴹ∙ᴬ x) ≈ (n ᴹ∙ᴬ y)
+
+-- derived form: iterated action, satisfying congruence
+
+    _ᴹ⋆ᴬ_ : List M → A → A
+    ms ᴹ⋆ᴬ a = foldr _ᴹ∙ᴬ_ a ms
+
+    ⋆-cong : Pointwise _≈ᴹ_ ms ns → x ≈ y → (ms ᴹ⋆ᴬ x) ≈ (ns ᴹ⋆ᴬ y)
+    ⋆-cong []            x≈y = x≈y
+    ⋆-cong (m≈n ∷ ms≋ns) x≈y = ∙-cong m≈n (⋆-cong ms≋ns x≈y)
+
+    ⋆-act-cong : Reflexive _≈ᴹ_ →
+                 ∀ ms ns → x ≈ y → ((ms ++ ns) ᴹ⋆ᴬ x) ≈ (ms ᴹ⋆ᴬ ns ᴹ⋆ᴬ y)
+    ⋆-act-cong refl []       ns x≈y = ⋆-cong {ns = ns} (Pointwise.refl refl) x≈y
+    ⋆-act-cong refl (m ∷ ms) ns x≈y = ∙-cong refl (⋆-act-cong refl ms ns x≈y)
+
+    []-act-cong : x ≈ y → ([] ᴹ⋆ᴬ x) ≈ y
+    []-act-cong = id
+
+  record RawRightAction : Set (a ⊔ r ⊔ c ⊔ ℓ)  where
+    infixl 5 _ᴬ∙ᴹ_ _ᴬ⋆ᴹ_
+
+    field
+      _ᴬ∙ᴹ_ : A → M → A
+      ∙-cong : x ≈ y → m ≈ᴹ n → (x ᴬ∙ᴹ m) ≈ (y ᴬ∙ᴹ n)
+
+-- derived form: iterated action, satisfying congruence
+
+    _ᴬ⋆ᴹ_ : A → List M → A
+    a ᴬ⋆ᴹ ms = foldl _ᴬ∙ᴹ_ a ms
+
+    ⋆-cong : x ≈ y → Pointwise _≈ᴹ_ ms ns → (x ᴬ⋆ᴹ ms) ≈ (y ᴬ⋆ᴹ ns)
+    ⋆-cong x≈y []            = x≈y
+    ⋆-cong x≈y (m≈n ∷ ms≋ns) = ⋆-cong (∙-cong x≈y m≈n) ms≋ns
+
+    ⋆-act-cong : Reflexive _≈ᴹ_ →
+                 x ≈ y → ∀ ms ns → (x ᴬ⋆ᴹ (ms ++ ns)) ≈ (y ᴬ⋆ᴹ ms ᴬ⋆ᴹ ns)
+    ⋆-act-cong refl x≈y []       ns = ⋆-cong {ns = ns} x≈y (Pointwise.refl refl)
+    ⋆-act-cong refl x≈y (m ∷ ms) ns = ⋆-act-cong refl (∙-cong x≈y refl) ms ns
+
+    []-act-cong : x ≈ y → (x ᴬ⋆ᴹ []) ≈ y
+    []-act-cong = id
+
+
+------------------------------------------------------------------------
+-- Definition: indexed over an underlying raw action
+
+module _ (M : Monoid c ℓ) (A : Setoid a r) where
+
+  private
+
+    open module M = Monoid M using () renaming (Carrier to M)
+    open module A = Setoid A using (_≈_) renaming (Carrier to A)
+    open ≋ M.setoid using (_≋_; [] ; _∷_)
+
+    variable
+      x y z : A.Carrier
+      m n p q : M.Carrier
+      ms ns ps qs : List M.Carrier
```

Comment:
Hmm this is the first attempt to have `variable` declarations mid-file. My opinion is that we shouldn't go here as it makes it much harder to parse a file.

Reply:
I'd like to push back on this, but if it's really a first use, perhaps I'm on shakier ground, but the pragmatic/ux gain seemed clear enough to me as a developer...

... but perhaps not (to you) as a reader/consumer of the module? 

Hmmm...

Reply:
But it's possible that the suggested breakup of this into much smaller modules might in fact lift such declarations to the top-level...?

Reply:
So the more non-local information you have to hold in your head at one time, the harder it is to understand a definition. When you have a collapse, and you have to go back and check everything to refresh it in your head, I find it much harder to build up the mental model again when I have to find and integrate non-local information in more places...

I guess it's exactly the same argument I have about having deeply nested modules within a file...

Reply:
> But it's possible that the suggested breakup of this into much smaller modules might in fact lift such declarations to the top-level...?

In which case, fantastic!

Reply:
Couldn't quite get it to work out :-(